lib/dotcom_web/live/trip_planner.ex

```diff
@@ -109,26 +122,75 @@ defmodule DotcomWeb.Live.TripPlanner do
     {:noreply, socket}
   end
 
-  # Selected from list of popular locations
-  defp location_props(%{"stop_id" => stop} = props) when is_binary(stop) do
-    Map.take(props, ["name", "latitude", "longitude", "stop_id"])
+  defp update_from_pin(socket, %{
+         "from" => %{"longitude" => from_longitude, "latitude" => from_latitude}
+       }) do
+    update_pin_in_socket(socket, [from_longitude, from_latitude], 0)
+  end
+
+  defp update_to_pin(socket, %{"to" => %{"longitude" => to_longitude, "latitude" => to_latitude}}) do
+    update_pin_in_socket(socket, [to_longitude, to_latitude], 1)
+  end
+
+  defp update_to_pin(socket, _params) do
+    socket
+  end
+
+  defp update_pin_in_socket(socket, [longitude, latitude], index)
+       when longitude != "" and latitude != "" do
+    pins =
+      place_pin(
+        socket.assigns.pins,
+        [String.to_float(longitude), String.to_float(latitude)],
+        index
+      )
+
+    socket |> assign(:pins, pins)
   end
 
-  # GTFS stop
-  defp location_props(%{"stop" => stop}) when is_map(stop) do
-    Map.take(stop, ["name", "latitude", "longitude"])
-    |> Map.put("stop_id", stop["id"])
+  defp update_pin_in_socket(socket, [longitude, latitude], index)
+       when longitude == "" or latitude == "" do
+    pins = remove_pin(socket.assigns.pins, index)
+
+    socket |> assign(:pins, pins)
+  end
+
+  defp update_pin_in_socket(socket, _coordinates, _index) do
+    socket
+  end
+
+  defp place_pin([], pin, 0) do
```

Comment:
They aren't 'magic' values. They are indexes in an array of length 2. The benefit is that we are possibly going to add more than a from and a to and have different waypoints like you can do with google maps. This makes it so that it takes an arbitrary number of pins.

Reply:
And, it's not just re-implementing `replace_at` or `insert_at`. Taken as an example:

```
defp place_pin([], pin, 1)
  [[], pin]
end
```

We need an empty pin so that the second pin gets the `B` rather than the `A`.

Reply:
And, we will add tests once the live view is finished.

Reply:
> And, it's not just re-implementing `replace_at` or `insert_at`. 

Touché

> The benefit is that we are possibly going to add more than a from and a to and have different waypoints like you can do with google maps

I would prefer to do that when we do that, rather than trying to build that framework into the live component now. That way we keep the code simple for now. Also, if we ever do need to support more than two pins, there's a decent chance that we won't implement it under the hood this way anyway (Skate detours did need to support variable numbers of waypoints, and we still found it helpful to have explicit start and end points).

And if we don't ever need to support more than two pins, then we keep the code nice and simple.

I added a suggested refactor in a different comment that I think addresses my concerns, and as a side-benefit, it dramatically reduces the line-of-code count. Lemme know what you think!